yo1k/tic_tac_toe/core/game.py

```diff
@@ -183,15 +186,26 @@ def __init__(self, action_queues: Sequence[ActionQueue]):
         self.__action_queues = action_queues
 
     def advance(self, state: State):
-        action = self.__action_queues[state.turn()].next()
-        if action is None:
-            return None
-        elif action.start is True:
-            Logic.__start(state)
-        elif action.surrender is True:
-            Logic.__surrender(state)
-        elif action.occupy is not None:
-            Logic.__occupy(state, action.occupy)
+        if state.phase is Phase.BEGINNING \
+                or state.phase is Phase.OUTROUND:
+            for player_idx in state.required_ready.copy():
```

Comment:
Do you need to copy the set because modifying it in `Logic.__ready` is not allowed concurrently with iterating over it via an iterator? In Java iterators allow removing elements from a collection while iterating (via the iterator itself, not directly via the collection), it's unfortunate that Python's API does not allow the same (I googled and checked the docs).

Reply:
Done.
> Do you need to copy the set because modifying it in `Logic.__ready` is not allowed concurrently with iterating over it via an iterator?

Yes, that is the reason.